### PR TITLE
[BEAM-11538] Add a Deque Encoder

### PR DESCRIPTION
**Pull request by Fokko**
**Please** add a meaningful description for your change here.

A Deque is in certain situations faster than a List: https://stackoverflow.com/questions/6163166/why-is-arraydeque-better-than-linkedlist. When writing a custom CombineFn, appending is something that we often do, therefore I think it is a nice addition to the default encoders.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`). R: @iemejia 
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Fokko** [comment]
@iemejia The CI is green now. What do you think of the Deque encoder?

**Comment by iemejia** [comment]
@Fokko can you please run `git pull origin master --rebase` it looks weird and I cannot see the changes here (and locally), but with this I can.

I was wondering why this was not done in the past just for not overspecializing? any ideas @kennknowles?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/DequeCoder.java
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/DequeCoder.java:73]: `DequeCoder.elemCoder` or better to even not reference private variables and say "Coder for elements of `DequeCoder` must be determistic".

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/DequeCoderTest.java
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/DequeCoderTest.java:48]: Prefer `ImmutableList.of` since this is not converting an array to a list, but just building a list.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/DequeCoderTest.java
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/DequeCoderTest.java:51]: It would be good to have some tests where the element type is not integer. For example for the case where the inner type is arrays or some other nested structure.

**Review by kennknowles — COMMENTED** [review]
It makes sense to me. I don't know why we do not have coders for a wider variety of standard structures. I guess because the idea is that elements should usually be very small and usually you have to observe the whole element. But this is a very good addition. Normally new features are experimental for a little while. In this case I think it would be better to skip that step. So then just more and more testing will be helpful.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/DequeCoder.java
Comment (Fokko) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/DequeCoder.java:73]: I've copied this from the `SetCoder`, but that's a good point!

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/DequeCoderTest.java
Comment (Fokko) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/coders/DequeCoderTest.java:48]: Good one, updated.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/DequeCoderTest.java
Comment (Fokko) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/coders/DequeCoderTest.java:51]: Sure, I've added the following test:
```java
  @Test
  public void encodeDequeWithList() throws Exception {
    DequeCoder<List<Long>> OBJECT_TEST_CODER = DequeCoder.of(ListCoder.of(VarLongCoder.of()));

    CoderProperties.coderDecodeEncodeContentsEqual(
        OBJECT_TEST_CODER,
        new ArrayDeque<>(
            ImmutableList.of(
                ImmutableList.of(18L, 15L), ImmutableList.of(19L, 25L), ImmutableList.of(22L))));
  }
```
Let me know if there are any other cases that you would like to see.

**Comment by Fokko** [comment]
Thanks @kennknowles for the review. I agree that it would be nice to have more standard Java structures in there by default. I had a simple combiner, where I changed from a List to a Deque, but it failed to run with a very hard-to-understand error message and took me quite a while to figure out what was going on.

**Comment by iemejia** [comment]
Seems checkstyle is not happy about the format @Fokko (You can test this locally with `/gradlew :sdks:java:core:checkstyleTest`, I suppose after fixing this it should be good to merge but let's wait to see if @kennknowles has some additional comments.

**Comment by iemejia** [comment]
checkstyle unhappy again

**Comment by Fokko** [comment]
🙀 I've fixed the import and ran the Spotless and Checkstyle.. twice!

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/DequeCoder.java
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/DequeCoder.java:55]: I don't believe this one is tested.

**Review by kennknowles — APPROVED** [review]
I want to be clear that I am really quite happy with this, but just that coders are particularly sensitive to changes.

I have a couple extra thoughts:

 - Basically I am reviewing by browsing `CoderProperties` and seeing which ones are not checked. It makes me realize that they are not very usable and it would be better if I had made it more like Guava's `EqualsTester` where it always performs roughly the same checks.
 - Coders also can be registered so they have a good encoding to portable protos. That is here: https://github.com/apache/beam/blob/3bb232fb098700de408f574585dfe74bbaff7230/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ModelCoderRegistrar.java#L48 (it is a bit of an odd architecture for legacy reasons, but all the proto translation bits are in runners/core/construction). Otherwise I am not exactly sure what the proto representation of the coder is, actually, but I assume it would be a Java SDK only thing, which maybe makes sense. This can be done later, too.
 -  Last point is that if it were to be a well-defined name for a well-defined encoding, we would maybe discuss on dev list and gather opinions from Python and Go. This coder seems useful enough that I could see it gaining usage that way.  This can be done later, too.

But LGTM with recommendation to really *really* test it a lot and if some other committer checks on this and thinks it is good they can merge. No need to wait for me. I sometimes have trouble with bandwidth for reviews these days.

**Comment by kennknowles** [comment]
Also not to anyone who merges it: don't merge all the fixup commits!

**Comment by iemejia** [comment]
Can you please rebase @Fokko and add the proposed test

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/DequeCoder.java
Comment (Fokko) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/DequeCoder.java:55]: I'm actually looking into this, but it isn't that trivial. This method is only tested for the `MapCoder` and `ListCoder`. My startingpoint was the tests of the `SetCoder`, where this wasn't covered. The issue with the Deque is that it do

`LinkedList` extends `AbstractSequentialList` which extends `AbstractList` which does override `equals` and `hashCode` - so the implementation is not inherited from `Object`. `ArrayDeque`, on the other hand, really doesn't inherit anything other implementation as far as I can see. Its direct superclass (`AbstractCollection`) doesn't override them. So the test is failing for the Deque since the objects aren't equal (it only tests `lhs == rhs`, as this is the equals implementation of Object.

**Comment by Fokko** [comment]
I've added the missing test :) It will check if the content of the iterable is equals.

**Review by iemejia — COMMENTED** [review]
Thanks for finding a way to test this and for the contribution @Fokko

It seems spotless is not happy about this, can you please fix that by running `./gradlew :sdks:java:core:spotlessApply` and do a `git pull origin master --rebase` to bring all your commits on top of master. After that can you please squash all of the commits into a single one and I will merge it afterwards. Thanks again!

**Comment by Fokko** [comment]
@iemejia Nicely squashed :)

**Review by iemejia — APPROVED** [review]
LGTM Thanks @Fokko, nice addition!

**Pull request merged**
